## Homework 9: Text Classification with Fine-Tuned BERT

### Due: Midnight on November 5th (with 2-hour grace period) — Worth 85 points

In this final homework, we’ll explore **fine-tuning a pre-trained Transformer model (BERT)** for text classification using the **IMDB Movie Review** dataset. You’ll begin with a working baseline notebook and then conduct a series of controlled experiments to understand how data size, context length, and model architecture affect performance.

You’ll complete three problems:

* **Problem 1:** Evaluate how **sequence length** and **learning rate** jointly influence validation loss and generalization.
* **Problem 2:** Measure how **training data size** affects both model performance and total training time.
* **Problem 3:** Compare **two additional models** from the BERT family to analyze the trade-offs between model size and accuracy on this dataset.

In each problem, you’ll report your key metrics, summarize what you observed, and reflect on what you learned.

> **Note:** This homework was developed and tested on **Google Colab**, due to version conflicts when running locally. It is **strongly recommended** that you complete your work on Colab as well.

There are 6 problems, each worth 14 points, and you get one point free if you complete the entire homework.


In [3]:
# Install once per new Colab runtime
%pip -q install -U keras keras-hub tensorflow tensorflow-text datasets evaluate

In [4]:

import os
os.environ["KERAS_BACKEND"] = "tensorflow"

import time
import random
import numpy as np
import keras
import keras_hub as kh
import evaluate
from datasets import load_dataset, Dataset, Features, Value, ClassLabel

from keras import mixed_precision                    # generally faster
mixed_precision.set_global_policy("mixed_float16")

### Here is where you can set global hyperparameters for this homework

In [5]:
# ---------------- Config ----------------
SEED        = 42
MAX_LEN     = 128
EPOCHS      = 3
BATCH       = 32
EVAL_BATCH  = 64
SUBSET_FRAC = 0.25   # <-- 0.25 to train and test on 25% of whole dataset during development;  set to 1.0 for full dataset

keras.utils.set_random_seed(SEED)




### Load and Preprocess the IMDB Movie Review Dataset

In [6]:
# ---- Load IMDb (raw), join train+test ----
imdb   = load_dataset("imdb")
texts  = list(imdb["train"]["text"]) + list(imdb["test"]["text"])
labels = np.array(list(imdb["train"]["label"]) + list(imdb["test"]["label"]), dtype="int32")

# ---- Build DS with explicit features (label=ClassLabel) ----
features = Features({"text": Value("string"),
                     "label": ClassLabel(num_classes=2, names=["NEG","POS"])})
all_ds = Dataset.from_dict({"text": texts, "label": labels.tolist()}, features=features)

# ---- Optional: take a stratified subset of the FULL dataset ----
if 0.0 < SUBSET_FRAC < 1.0:
    sub = all_ds.train_test_split(train_size=SUBSET_FRAC, seed=SEED, stratify_by_column="label")
    ds_pool = sub["train"]
else:
    ds_pool = all_ds

# ---- Stratified 80/10/10 split on the (possibly smaller) pool ----
# First: 80/20 train+val pool / test
splits = ds_pool.train_test_split(test_size=0.20, seed=SEED, stratify_by_column="label")
train_val_pool, test_ds = splits["train"], splits["test"]
# Then: carve 10% of full (i.e., 0.125 of the 80% pool) as validation
splits2 = train_val_pool.train_test_split(test_size=0.125, seed=SEED, stratify_by_column="label")
train_ds, val_ds = splits2["train"], splits2["test"]

# ---- Numpy arrays for Keras fit/predict ----
X_tr = np.array(train_ds["text"], dtype=object); y_tr = np.array(train_ds["label"], dtype="int32")
X_va = np.array(val_ds["text"],   dtype=object); y_va = np.array(val_ds["label"],   dtype="int32")
X_te = np.array(test_ds["text"],  dtype=object); y_te = np.array(test_ds["label"],  dtype="int32")

# ---- Quick summary ----
def _counts(ds):
    arr = np.array(ds["label"], dtype=int)
    return len(arr), np.bincount(arr, minlength=2).tolist()
print(f"Pool after SUBSET_FRAC={SUBSET_FRAC}: {len(ds_pool)} (of {len(all_ds)})")
print("Train:", _counts(train_ds), " Val:", _counts(val_ds), " Test:", _counts(test_ds))


/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


README.md: 0.00B [00:00, ?B/s]

plain_text/train-00000-of-00001.parquet:   0%|          | 0.00/21.0M [00:00<?, ?B/s]

plain_text/test-00000-of-00001.parquet:   0%|          | 0.00/20.5M [00:00<?, ?B/s]

plain_text/unsupervised-00000-of-00001.p(…):   0%|          | 0.00/42.0M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/25000 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/25000 [00:00<?, ? examples/s]

Generating unsupervised split:   0%|          | 0/50000 [00:00<?, ? examples/s]

Pool after SUBSET_FRAC=0.25: 12500 (of 50000)
Train: (8750, [4375, 4375])  Val: (1250, [625, 625])  Test: (2500, [1250, 1250])


### Build and train a baseline Distil-Bert Text Classifier

In [7]:
# ---- Keras Hub preprocessor + classifier ----
preproc = kh.models.DistilBertTextClassifierPreprocessor.from_preset(
    "distil_bert_base_en_uncased", sequence_length=MAX_LEN
)
model = kh.models.DistilBertTextClassifier.from_preset(
    "distil_bert_base_en_uncased", num_classes=2, preprocessor=preproc
)

model.compile(
    optimizer=keras.optimizers.Adam(1e-5),
    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
    metrics=[keras.metrics.SparseCategoricalAccuracy(name="acc")],
)

start = time.time()

# ---- Train with early stopping (restore best val weights) ----
cb = [keras.callbacks.EarlyStopping(monitor="val_loss", patience=2, restore_best_weights=True)]
history = model.fit(
    X_tr, y_tr,
    validation_data=(X_va, y_va),
    epochs=EPOCHS,
    batch_size=BATCH,
    callbacks=cb,
    verbose=1,
)

# ---- Evaluate (accuracy + F1 via `evaluate`) ----
logits = model.predict(X_te, batch_size=EVAL_BATCH, verbose=0)
y_pred = logits.argmax(axis=-1)

acc_metric = evaluate.load("accuracy")
f1_metric  = evaluate.load("f1")
acc = acc_metric.compute(predictions=y_pred, references=y_te)["accuracy"]
f1  = f1_metric.compute(predictions=y_pred, references=y_te)["f1"]

# Tiny confusion matrix helper (no sklearn needed)
def confusion_matrix_np(y_true, y_pred, num_classes=2):
    cm = np.zeros((num_classes, num_classes), dtype=int)
    for t, p in zip(y_true, y_pred):
        cm[t, p] += 1
    return cm

print(f"\nValidation acc (best epoch): {history.history['val_acc'][np.argmin(history.history['val_loss'])]:.3f}")
print(f"\nTest accuracy: {acc:.3f}   Test F1: {f1:.3f}")
print("\nConfusion matrix:\n", confusion_matrix_np(y_te, y_pred))

end = time.time() - start
print("\nElapsed time:", time.strftime("%H:%M:%S", time.gmtime(end)))

Epoch 1/3
274/274 ━━━━━━━━━━━━━━━━━━━━ 122s 249ms/step - acc: 0.7824 - loss: 0.4530 - val_acc: 0.8376 - val_loss: 0.3447
Epoch 2/3
274/274 ━━━━━━━━━━━━━━━━━━━━ 10s 35ms/step - acc: 0.8789 - loss: 0.2897 - val_acc: 0.8584 - val_loss: 0.3403
Epoch 3/3
274/274 ━━━━━━━━━━━━━━━━━━━━ 10s 34ms/step - acc: 0.9159 - loss: 0.2207 - val_acc: 0.8600 - val_loss: 0.3555



Validation acc (best epoch): 0.858

Test accuracy: 0.854   Test F1: 0.851

Confusion matrix:
 [[1096  154]
 [ 210 1040]]

Elapsed time: 00:02:48


# Problem 1 — Mini sweep: context length × learning rate (6 runs)

In this problem we'll see how much **context length** (`MAX_LEN`) helps, and how sensitive fine-tuning is to **learning rate**—without running a huge grid.

## Setup (keep these fixed)

* `SUBSET_FRAC = 0.25`               # use only this percentage of the whole dataset
* `EPOCHS = 3`
* `BATCH = 32` (but see note for 256 below)
* **EarlyStopping** with `restore_best_weights=True`
* Same random `SEED` for all runs
* Same data split for all runs (don’t reshuffle between runs)

### Run these 6 configurations

**For each** `MAX_LEN ∈ {128, 256, 512}`, try **two** learning rates:

* **MAX_LEN = 128**

  * `(LR = 2e-5, BATCH = 32)` – healthy default for shorter contexts.
  * `(LR = 1e-5, BATCH = 32)` – conservative LR; often a touch stabler.

* **MAX_LEN = 256**

  * `(LR = 1e-5, BATCH = 16)` – longer context → lower batch.
  * `(LR = 7.5e-6, BATCH = 16)` – even steadier if loss is noisy.

* **MAX_LEN = 512**  *(heavier quadratic attention cost)*

  * `(LR = 7.5e-6, BATCH = 8)` – safe starting point.
  * `(LR = 5e-6, BATCH = 8)` – extra caution for stability.

**If you hit an Out Of Memory error:**

* At **256** with `BATCH = 16`, drop to `BATCH = 8`.
* At **512** with `BATCH = 8`, drop to `BATCH = 4`.


Then answer the graded questions.


In [8]:
# Your code here; add as many cells as you need

import time, numpy as np, keras, keras_hub as kh
from keras import callbacks as kcb

def best_val_acc_at_min_loss(history):
    """Return (val_acc_at_min_val_loss, best_epoch_index) from Keras History."""
    vl = history.history["val_loss"]
    va = history.history.get("val_acc") or history.history.get("val_accuracy")
    i = int(np.argmin(vl))
    return float(va[i]), i

def run_one(MAX_LEN, LR, BATCH, epochs=EPOCHS):
    preproc = kh.models.DistilBertTextClassifierPreprocessor.from_preset(
        "distil_bert_base_en_uncased", sequence_length=MAX_LEN
    )
    model = kh.models.DistilBertTextClassifier.from_preset(
        "distil_bert_base_en_uncased", num_classes=2, preprocessor=preproc
    )
    model.compile(
        optimizer=keras.optimizers.Adam(LR),
        loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
        metrics=[keras.metrics.SparseCategoricalAccuracy(name="acc")],
    )
    es = kcb.EarlyStopping(monitor="val_loss", patience=2, restore_best_weights=True)
    start = time.time()
    try:
        hist = model.fit(
            X_tr, y_tr,
            validation_data=(X_va, y_va),
            epochs=epochs,
            batch_size=BATCH,
            callbacks=[es],
            verbose=1,
        )
    except Exception as e:
        # Friendly OOM fallback: try halving batch once
        if "OOM" in str(e) or "resource exhausted" in str(e).lower():
            alt_batch = max(1, BATCH // 2)
            print(f"[WARN] OOM at batch={BATCH}. Retrying with batch={alt_batch} …")
            hist = model.fit(
                X_tr, y_tr,
                validation_data=(X_va, y_va),
                epochs=epochs,
                batch_size=alt_batch,
                callbacks=[es],
                verbose=1,
            )
        else:
            raise
    wall = time.time() - start
    val_acc_at_min, best_ep = best_val_acc_at_min_loss(hist)
    return {
        "MAX_LEN": MAX_LEN,
        "LR": LR,
        "BATCH": BATCH,
        "val_acc_at_min_val_loss": val_acc_at_min,
        "best_epoch_idx": best_ep,    # 0-based
        "epochs_trained": len(hist.history["loss"]),
        "wall_sec": wall,
    }

sweep = [
    # MAX_LEN=128
    {"MAX_LEN":128, "LR":2e-5,  "BATCH":32},
    {"MAX_LEN":128, "LR":1e-5,  "BATCH":32},
    # MAX_LEN=256
    {"MAX_LEN":256, "LR":1e-5,  "BATCH":16},
    {"MAX_LEN":256, "LR":7.5e-6,"BATCH":16},
    # MAX_LEN=512
    {"MAX_LEN":512, "LR":7.5e-6,"BATCH":8},
    {"MAX_LEN":512, "LR":5e-6,  "BATCH":8},
]

results = []
for cfg in sweep:
    print(f"\n=== Run: MAX_LEN={cfg['MAX_LEN']} | LR={cfg['LR']} | BATCH={cfg['BATCH']} ===")
    out = run_one(**cfg, epochs=EPOCHS)
    results.append(out)
    print("→ val_acc@min_val_loss = {:.4f} | best_epoch={} | time={:,.1f}s".format(
        out["val_acc_at_min_val_loss"], out["best_epoch_idx"]+1, out["wall_sec"]
    ))

# Pretty summary table
print("\nSummary (sorted by val_acc@min_val_loss desc):")
results_sorted = sorted(results, key=lambda r: r["val_acc_at_min_val_loss"], reverse=True)
for r in results_sorted:
    print("MAX_LEN={:<3}  LR={:<8g}  BATCH={:<3}  val_acc@min={:.4f}  best_ep={:<2}  time={:,.1f}s".format(
        r["MAX_LEN"], r["LR"], r["BATCH"], r["val_acc_at_min_val_loss"], r["best_epoch_idx"]+1, r["wall_sec"]
    ))

# a1a = validation accuracy at min validation loss for the BEST configuration
best = results_sorted[0]
a1a = float(best["val_acc_at_min_val_loss"])
a1a



=== Run: MAX_LEN=128 | LR=2e-05 | BATCH=32 ===
Epoch 1/3
274/274 ━━━━━━━━━━━━━━━━━━━━ 86s 159ms/step - acc: 0.8009 - loss: 0.4151 - val_acc: 0.8480 - val_loss: 0.3463
Epoch 2/3
274/274 ━━━━━━━━━━━━━━━━━━━━ 10s 35ms/step - acc: 0.9024 - loss: 0.2484 - val_acc: 0.8360 - val_loss: 0.3904
Epoch 3/3
274/274 ━━━━━━━━━━━━━━━━━━━━ 10s 34ms/step - acc: 0.9369 - loss: 0.1706 - val_acc: 0.8504 - val_loss: 0.3788
→ val_acc@min_val_loss = 0.8480 | best_epoch=1 | time=106.5s

=== Run: MAX_LEN=128 | LR=1e-05 | BATCH=32 ===
Epoch 1/3
274/274 ━━━━━━━━━━━━━━━━━━━━ 88s 160ms/step - acc: 0.7817 - loss: 0.4544 - val_acc: 0.8528 - val_loss: 0.3442
Epoch 2/3
274/274 ━━━━━━━━━━━━━━━━━━━━ 10s 35ms/step - acc: 0.8802 - loss: 0.2905 - val_acc: 0.8568 - val_loss: 0.3484
Epoch 3/3
274/274 ━━━━━━━━━━━━━━━━━━━━ 10s 35ms/step - acc: 0.9150 - loss: 0.2203 - val_acc: 0.8584 - val_loss: 0.3544
→ val_acc@min_val_loss = 0.8528 | best_epoch=1 | time=108.5s

=== Run: MAX_LEN=256 | LR=1e-05 | BATCH=16 ===
Epoch 1/3
547/547 

0.9200000166893005

### Graded Questions

In [9]:
# Set a1a to the validation accuracy at min validation loss for your best configuration found in this problem

a1a = 0.9200             # Replace 0.0 with your answer

In [10]:
# Graded Answer
# DO NOT change this cell in any way

print(f'a1a = {a1a:.4f}')

a1a = 0.9200


#### Question a1b:

* Does **more context** (128 → 256 → 512) consistently help?
* How much effect did the learning rate have on the validation accuracy?


#### Your Answer Here:
validation accuracy rose steadily from 85 % to 90 % to 92 % as the context length increased.
The largest jump came between 128 and 256 tokens and 512 brought another small but clear gain, suggesting that longer context does capture more sentiment cues though with much higher compute cost.

The effect was modest but visible. Within each context length, the smaller learning rate was slightly more stable and gave 0.3–0.7 pp higher validation accuracy, especially for longer sequences where gradients can be noisier. At MAX_LEN = 128, 2e-5 and 1e-5 were nearly tied; for 256 and 512, the lower LR yielded the best convergence and avoided loss spikes.

## Problem 2 — How much data is enough?

In this problem, you’ll investigate how model performance scales with dataset size.

**Setup.**
Use the best `MAX_LEN` and `LR` values you found in **Problem 1**.

**What to do:**

1. For each value of `SUBSET_FRAC ∈ {0.25, 0.50, 0.75, 1.00}`, train your model once and observe the displayed performance metrics.
2. Answer the discussion question below.




In [11]:
# Your code here; add as many cells as you need

SEED = 42
MAX_LEN = 512
LR = 5e-6
EPOCHS = 3
BATCH = 8
EVAL_BATCH = 64
SUBSET_FRACS = [0.25, 0.50, 0.75, 1.00]

def load_split(sub_frac):
    imdb   = load_dataset("imdb")
    texts  = list(imdb["train"]["text"]) + list(imdb["test"]["text"])
    labels = np.array(list(imdb["train"]["label"]) + list(imdb["test"]["label"]), dtype="int32")

    feats = Features({"text": Value("string"),
                      "label": ClassLabel(num_classes=2, names=["NEG","POS"])})
    all_ds = Dataset.from_dict({"text": texts, "label": labels.tolist()}, features=feats)

    ds_pool = (
        all_ds.train_test_split(train_size=sub_frac, seed=SEED, stratify_by_column="label")["train"]
        if 0.0 < sub_frac < 1.0 else all_ds
    )

    s1 = ds_pool.train_test_split(test_size=0.20, seed=SEED, stratify_by_column="label")
    train_val_pool, test_ds = s1["train"], s1["test"]
    s2 = train_val_pool.train_test_split(test_size=0.125, seed=SEED, stratify_by_column="label")
    train_ds, val_ds = s2["train"], s2["test"]

    X_tr, y_tr = np.array(train_ds["text"], dtype=object), np.array(train_ds["label"], dtype="int32")
    X_va, y_va = np.array(val_ds["text"], dtype=object), np.array(val_ds["label"], dtype="int32")
    X_te, y_te = np.array(test_ds["text"], dtype=object), np.array(test_ds["label"], dtype="int32")
    return X_tr, y_tr, X_va, y_va, X_te, y_te, len(ds_pool)

def run_one(sub_frac):
    X_tr, y_tr, X_va, y_va, X_te, y_te, n = load_split(sub_frac)
    preproc = kh.models.DistilBertTextClassifierPreprocessor.from_preset(
        "distil_bert_base_en_uncased", sequence_length=MAX_LEN
    )
    model = kh.models.DistilBertTextClassifier.from_preset(
        "distil_bert_base_en_uncased", num_classes=2, preprocessor=preproc
    )
    model.compile(
        optimizer=keras.optimizers.Adam(LR),
        loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
        metrics=[keras.metrics.SparseCategoricalAccuracy(name="acc")],
    )
    es = kcb.EarlyStopping(monitor="val_loss", patience=2, restore_best_weights=True)
    start = time.time()
    hist = model.fit(X_tr, y_tr,
                     validation_data=(X_va, y_va),
                     epochs=EPOCHS, batch_size=BATCH,
                     callbacks=[es], verbose=1)
    val_acc = hist.history["val_acc"][np.argmin(hist.history["val_loss"])]
    logits = model.predict(X_te, batch_size=EVAL_BATCH, verbose=0)
    y_pred = logits.argmax(axis=-1)
    acc_metric = evaluate.load("accuracy")
    f1_metric  = evaluate.load("f1")
    acc = acc_metric.compute(predictions=y_pred, references=y_te)["accuracy"]
    f1  = f1_metric.compute(predictions=y_pred, references=y_te)["f1"]
    dur = time.time() - start
    return {"SUBSET_FRAC":sub_frac, "n_samples":n, "val_acc@min":val_acc,
            "test_acc":acc, "test_f1":f1, "time_sec":dur}

results2 = []
for f in SUBSET_FRACS:
    print(f"\n=== Run: SUBSET_FRAC={f} ===")
    out = run_one(f)
    results2.append(out)
    print("→ val_acc@min={:.4f}  test_acc={:.4f}  f1={:.4f}  time={:,.1f}s".format(
        out["val_acc@min"], out["test_acc"], out["test_f1"], out["time_sec"])
    )

print("\nSummary:")
for r in results2:
    print("SUBSET={:<4.2f}  n={:<6}  val@min={:.4f}  test_acc={:.4f}  f1={:.4f}  time={:,.1f}s".format(
        r["SUBSET_FRAC"], r["n_samples"], r["val_acc@min"], r["test_acc"], r["test_f1"], r["time_sec"])
    )



=== Run: SUBSET_FRAC=0.25 ===
Epoch 1/3
1094/1094 ━━━━━━━━━━━━━━━━━━━━ 109s 60ms/step - acc: 0.8466 - loss: 0.3454 - val_acc: 0.9064 - val_loss: 0.2295
Epoch 2/3
1094/1094 ━━━━━━━━━━━━━━━━━━━━ 34s 31ms/step - acc: 0.9275 - loss: 0.1948 - val_acc: 0.9152 - val_loss: 0.2322
Epoch 3/3
1094/1094 ━━━━━━━━━━━━━━━━━━━━ 33s 30ms/step - acc: 0.9496 - loss: 0.1386 - val_acc: 0.9040 - val_loss: 0.2638
→ val_acc@min=0.9064  test_acc=0.9144  f1=0.9146  time=202.6s

=== Run: SUBSET_FRAC=0.5 ===
Epoch 1/3
2188/2188 ━━━━━━━━━━━━━━━━━━━━ 151s 50ms/step - acc: 0.8753 - loss: 0.2918 - val_acc: 0.9208 - val_loss: 0.1955
Epoch 2/3
2188/2188 ━━━━━━━━━━━━━━━━━━━━ 67s 30ms/step - acc: 0.9315 - loss: 0.1839 - val_acc: 0.9244 - val_loss: 0.1890
Epoch 3/3
2188/2188 ━━━━━━━━━━━━━━━━━━━━ 66s 30ms/step - acc: 0.9514 - loss: 0.1365 - val_acc: 0.9216 - val_loss: 0.2206
→ val_acc@min=0.9244  test_acc=0.9278  f1=0.9274  time=298.1s

=== Run: SUBSET_FRAC=0.75 ===
Epoch 1/3
3282/3282 ━━━━━━━━━━━━━━━━━━━━ 186s 44ms/step 

### Graded Questions

In [12]:
# Set a2a to the validation accuracy at min validation loss for your best configuration found in this problem
# (Yes, it is probably at 1.0!)

a2a = 0.9314            # Replace 0.0 with your answer

In [13]:
# Graded Answer
# DO NOT change this cell in any way

print(f'a2a = {a2a:.4f}')

a2a = 0.9314


#### Question a2b:

Summarize what you observed as dataset size increased. Given that validation metrics are typically reliable to only about two decimal places, do the performance gains justify using the entire dataset? What trade-offs between accuracy and computation time did you notice?

#### Your Answer Here:
As I used more data, the accuracy went up a bit but started to level off after about 75%. The full dataset gave the best score (around 93%), but the gain over smaller subsets was tiny. Training time, though, almost doubled. So using all the data isn’t really worth it unless you need that last fraction of a percent.

# Problem 3 — Model swap: speed vs. accuracy (why: capacity matters)

In this problem we will compare encoder-only backbones of different sizes.

**Setup.** Keep the best `MAX_LEN`, `LR`, and `SUBSET_FRAC` from Problems 1–2. Only change the model/preset:

* **DistilBERT** (current baseline)
* **BERT-base** (larger/usually stronger)

**How to switch (two lines each).**

* DistilBERT:

  ```python
  preproc = kh.models.DistilBertTextClassifierPreprocessor.from_preset("distil_bert_base_en_uncased", sequence_length=MAX_LEN)
  model  = kh.models.DistilBertTextClassifier.from_preset("distil_bert_base_en_uncased", num_classes=2, preprocessor=preproc)
  ```

* BERT-base:

  ```python
  preproc = kh.models.BertTextClassifierPreprocessor.from_preset("bert_base_en_uncased", sequence_length=MAX_LEN)
  model  = kh.models.BertTextClassifier.from_preset("bert_base_en_uncased", num_classes=2, preprocessor=preproc)
  ```

**What to do.**

1. Train/evaluate each model once with identical settings.
2. Observe the performance metrics for each.
3. Answer the graded questions.



In [14]:
# Your code here; add as many cells as you wish


import time, numpy as np, keras, keras_hub as kh, evaluate
from keras import callbacks as kcb

MAX_LEN = 512
LR      = 5e-6
EPOCHS  = 3
BATCH   = 8
EVAL_BATCH = 64

def run_kh_model(kind, preset):
    pre = kind["preproc"].from_preset(preset, sequence_length=MAX_LEN)
    mdl = kind["model"].from_preset(preset, num_classes=2, preprocessor=pre)
    mdl.compile(
        optimizer=keras.optimizers.Adam(LR),
        loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
        metrics=[keras.metrics.SparseCategoricalAccuracy(name="acc")],
    )
    es = kcb.EarlyStopping(monitor="val_loss", patience=2, restore_best_weights=True)
    t0 = time.time()
    hist = mdl.fit(X_tr, y_tr, validation_data=(X_va, y_va),
                   epochs=EPOCHS, batch_size=BATCH, callbacks=[es], verbose=1)
    wall = time.time() - t0

    val_idx = int(np.argmin(hist.history["val_loss"]))
    val_acc = float(hist.history["val_acc"][val_idx])

    logits = mdl.predict(X_te, batch_size=EVAL_BATCH, verbose=0)
    y_pred = logits.argmax(axis=-1)

    acc = evaluate.load("accuracy").compute(predictions=y_pred, references=y_te)["accuracy"]
    f1  = evaluate.load("f1").compute(predictions=y_pred, references=y_te)["f1"]
    return val_acc, acc, f1, wall

kinds = {
    "DistilBERT": {
        "preproc": kh.models.DistilBertTextClassifierPreprocessor,
        "model":   kh.models.DistilBertTextClassifier,
        "preset":  "distil_bert_base_en_uncased",
    },
    "BERT-base": {
        "preproc": kh.models.BertTextClassifierPreprocessor,
        "model":   kh.models.BertTextClassifier,
        "preset":  "bert_base_en_uncased",
    }
}

results3 = {}
for name, spec in kinds.items():
    print(f"\n=== {name} ===")
    vacc, tacc, f1, secs = run_kh_model(
        {"preproc": spec["preproc"], "model": spec["model"]}, spec["preset"]
    )
    results3[name] = dict(val_acc=vacc, test_acc=tacc, f1=f1, time_sec=secs)
    print(f"→ val@min={vacc:.4f}  test_acc={tacc:.4f}  f1={f1:.4f}  time={secs:,.1f}s")

# Set a3a = best validation accuracy at min val loss
best_name = max(results3, key=lambda k: results3[k]["val_acc"])
a3a = float(results3[best_name]["val_acc"])
print("\nBest model:", best_name, "| a3a =", a3a)
results3



=== DistilBERT ===
Epoch 1/3
1094/1094 ━━━━━━━━━━━━━━━━━━━━ 111s 61ms/step - acc: 0.8463 - loss: 0.3440 - val_acc: 0.9136 - val_loss: 0.2310
Epoch 2/3
1094/1094 ━━━━━━━━━━━━━━━━━━━━ 36s 32ms/step - acc: 0.9274 - loss: 0.1953 - val_acc: 0.9104 - val_loss: 0.2276
Epoch 3/3
1094/1094 ━━━━━━━━━━━━━━━━━━━━ 35s 31ms/step - acc: 0.9520 - loss: 0.1376 - val_acc: 0.9080 - val_loss: 0.2488
→ val@min=0.9104  test_acc=0.9156  f1=0.9162  time=182.5s

=== BERT-base ===


100%|██████████| 457/457 [00:00<00:00, 948kB/s]


100%|██████████| 761/761 [00:00<00:00, 1.43MB/s]


100%|██████████| 226k/226k [00:00<00:00, 2.16MB/s]


100%|██████████| 418M/418M [00:05<00:00, 84.5MB/s]


Epoch 1/3
1094/1094 ━━━━━━━━━━━━━━━━━━━━ 193s 106ms/step - acc: 0.8698 - loss: 0.3068 - val_acc: 0.9088 - val_loss: 0.2261
Epoch 2/3
1094/1094 ━━━━━━━━━━━━━━━━━━━━ 65s 59ms/step - acc: 0.9419 - loss: 0.1654 - val_acc: 0.9144 - val_loss: 0.2255
Epoch 3/3
1094/1094 ━━━━━━━━━━━━━━━━━━━━ 65s 59ms/step - acc: 0.9694 - loss: 0.0970 - val_acc: 0.9048 - val_loss: 0.2707
→ val@min=0.9144  test_acc=0.9152  f1=0.9174  time=324.9s

Best model: BERT-base | a3a = 0.9143999814987183


{'DistilBERT': {'val_acc': 0.9103999733924866,
  'test_acc': 0.9156,
  'f1': 0.9161700437028208,
  'time_sec': 182.5107421875},
 'BERT-base': {'val_acc': 0.9143999814987183,
  'test_acc': 0.9152,
  'f1': 0.9174454828660437,
  'time_sec': 324.9089639186859}}

### Graded Questions

In [15]:
# Set a1a to the validation accuracy at min validation loss for your best model found in this problem

a3a = 0.9144           # Replace 0.0 with your answer

In [16]:
# Graded Answer
# DO NOT change this cell in any way

print(f'a3a = {a3a:.4f}')

a3a = 0.9144


#### Question a3b:

**Answer briefly.**

* Which model gives the best **accuracy/F1**?
* Which is **fastest** per epoch?
* Given limited development time or compute resources, which model is the best **overall choice** and why?

#### Your Answer Here:

BERT-base gave slightly better accuracy and F1 than DistilBERT..
DistilBERT was much faster—about 180 s vs. 325 s per run.
If time or compute is limited, DistilBERT is the better all-around choice since it’s quicker and nearly as accurate.